### PR TITLE
[Notifications] Fix local sync notifications not being run

### PR DESCRIPTION
**Pull request by quaark**
Fixes - https://jira.iguazeng.com/browse/ML-3962

Separate sync and async notifications. Run sync notifications ONLY on client side in main thread. The async notifications can be run anywhere by sending them to the main event loop.

### CODE REVIEW

File: tests/utils/test_notifications.py [hunk 1]
```diff
@@ -353,6 +353,9 @@ def test_inverse_dependencies(
     )
 
     custom_notification_pusher.push("test-message", "info", [])
+    import time
+
+    time.sleep(3)
```
Comment (liranbg) [inline tests/utils/test_notifications.py:358]: mind adding some retry until succesfull with a 10 second timeout ? i think it will fail transiently

**Review by liranbg — CHANGES_REQUESTED** [review]
overall looking good.
I am missing the "why" here as I understand the "what". make sure to add on code / pr description so it will make sense in the future too.

File: tests/utils/test_notifications.py (hunk 1 above)
Comment (quaark) [reply tests/utils/test_notifications.py:358]: Ah this is a leftover. removing it

Comment (quaark) [reply tests/utils/test_notifications.py:358]: Ah this is a leftover, isn't needed. removing

**Pull request merged**